### PR TITLE
Remove all uses of `NodeId` in `ResolverOutputs`

### PR DESCRIPTION
**Pull request by marmeladema**
cc #50928 

r? @ecstatic-morse

### CODE REVIEW

File: src/librustc_resolve/lib.rs
Comment (ecstatic-morse) [inline src/librustc_resolve/lib.rs:1277]: `HashMap` implements `FromIterator`. Why not use `collect` for all these like the old code does?

File: src/librustc_middle/ty/mod.rs
Comment (ecstatic-morse) [inline src/librustc_middle/ty/mod.rs:128]: It does seem like all of these could plausibly be `DefId`. Did you find something during your experiments that wasn't a definition? If not, we should open a cleanup issue once this gets merged.

**Review by ecstatic-morse — CHANGES_REQUESTED** [review]
cc @ljedrz

Left a nit. I'll approve once it is fixed and CI is green. Thanks @marmeladema!

File: src/librustc_middle/ty/mod.rs
Comment (marmeladema) [reply src/librustc_middle/ty/mod.rs:128]: I've only tried for `trait_map` but yes, for this field, there is no def id, only hir id

File: src/librustc_resolve/lib.rs
Comment (marmeladema) [reply src/librustc_resolve/lib.rs]: Fixed!

File: src/librustc_hir/definitions.rs
Comment (ljedrz) [inline src/librustc_hir/definitions.rs:331]: `unwrap_or_else(|| panic!(x))` is equivalent to `expect(x)`

File: src/librustc_resolve/lib.rs
Comment (ljedrz) [inline src/librustc_resolve/lib.rs:1345]: nit: `cloned()` after `iter()` should allow us to drop the individual `clone`s in the `map`s; also applies to the few other instances below

File: src/librustc_hir/definitions.rs
Comment (Xanewok) [reply src/librustc_hir/definitions.rs:331]: I don't think you can lazily format with `expect`, though

Comment (ljedrz) [reply src/librustc_hir/definitions.rs:331]: Ah, you're right, I forgot about that :+1:.

Comment (marmeladema) [reply src/librustc_hir/definitions.rs:331]: Yeah that's why I did not use `expect` :-)

File: src/librustc_resolve/lib.rs
Comment (ecstatic-morse) [reply src/librustc_resolve/lib.rs:1345]: I don't care so much about `cloned` vs `clone`, but there do seem to be some calls to `clone` for `Copy` types, which [clippy will lint against](https://github.com/rust-lang/rust-clippy/pull/602). Only using `clone` when necessary makes it easier to spot places that do excessive heap allocation. Using `|(&k, &v)|` here, or preferably `Iterator::copied` would be preferable.

File: src/librustc_middle/ty/mod.rs
Comment (ecstatic-morse) [reply src/librustc_middle/ty/mod.rs:128]: @petrochenkov I see you've assigned yourself. Any recommendations here? Is there a reason that `trait_map` and `export_map` cannot use `DefId`s?

File: src/librustc_resolve/lib.rs
Comment (marmeladema) [reply src/librustc_resolve/lib.rs:1345]: I don't think it works in this case because `HashMap::iter` will create an iterator that yields tuples `(&k, &v)` and iirc `cloned()` will work only on references. I could probably clone the entire map before and call `.into_iter()` instead?

Comment (ecstatic-morse) [reply src/librustc_resolve/lib.rs:1345]: Ah that's right. In that case use the argument binding approach (`|(&k, &v)|`) for the `Copy` types. Cloning the entire map will cause the backing storage to be unnecessarily cloned as well.

Comment (marmeladema) [reply src/librustc_resolve/lib.rs]: Done! Thank you for the advises!

**Comment by marmeladema** [comment]
Rebased and fixed comments

File: src/librustc_middle/ty/mod.rs
Comment (marmeladema) [reply src/librustc_middle/ty/mod.rs:128]: Also because the global context uses `HirId`, using `DefId` here would mean another map clone/conversion in `librustc_middle`.

**Comment by ecstatic-morse** [comment]
Thanks @marmeladema! It's possible that someone with more knowledge of AST -> HIR lowering (@petrochenkov?) could find a solution to https://github.com/rust-lang/rust/pull/72402#discussion_r428368228. However, because this may unlock https://github.com/rust-lang/rust/issues/50928#issuecomment-631594659, I'm happy to approve in its present state.

@bors r+

File: src/librustc_middle/ty/mod.rs
Comment (petrochenkov) [reply src/librustc_middle/ty/mod.rs:128]: The zoo of `DefId`s, `LocalDefId`s and `HirId`s here is a bit surprising (all of these could be `LocalDefId`s (or `HirId`s)), but it looks like the choices are made to minimize conversions at use sites.

I guess the right thing to do here would be changing `DefId`s into `LocalDefId`s in `GlobalCtxt` which is the end destination of these tables.

Comment (marmeladema) [reply src/librustc_middle/ty/mod.rs:128]: I can definitely take a look at that in a follow up PR!

**Pull request merged**
